docs/001-ci-strategy.adoc

```diff
@@ -0,0 +1,101 @@
+
+== Context
+
+A continuous integration and continuous delivery platform (CI/CD) should help us to deliver high quality software to our users.
+In the past we have been heavily using https://github.com/features/actions[GitHub Actions] but we had our problems with performance and maintenance of the CI suite.
+Below is a short description on what we want to achieve and how.
+
+
+=== High Test Coverage and Assurance
+
+We want to have high test coverage due to obvious reason.
+A high coverage only makes sense if it is run frequently.
+To have viable tests run on the CI, they should run fast.
+Our role model Martin Fowler describes that tests on CI should not take more than https://martinfowler.com/articles/continuousIntegration.html#KeepTheBuildFast[10 minutes].
+
+[quote]
+For most projects, however, the XP guideline of a ten minute build is perfectly within reason. Most of our modern projects achieve this. It's worth putting in concentrated effort to make it happen, because every minute you reduce off the build time is a minute saved for each developer every time they commit. Since CI demands frequent commits, this adds up to a lot of time.
+
+He goes on with
+
+[quote]
+The trick is to balance the needs of bug finding and speed so that a good commit build is stable enough for other people to work on.
+Once the commit build is good then other people can work on the code with confidence. However, there are further, slower, tests that you can start to do. Additional machines can run further testing routines on the build that take longer to do.
+
+He proposes to make use of a two stage deployment pipeline:
+
+1. Stage 1: Run a build and unit tests with the database/io/network/etc completely stubbed out.
+2. Stage 2: Run larger scale interaction tests.
+
+The main goal of stage 1 is to achieve high certainty that others can start working on it.
+Only in stage 2 longer tests will run.
+
+The idea is to only run stage 2 tests on a regular interval and not on every commit.
+
+If we keep following our well-known `bors r+`-workflow, we can execute stage 2 tests only when merging into the `staging` branch, i.e. on or when `bors r+` is executed.
+
+The advantages of this two stage approach are:
+
+1. separation of test files: in order to be able to differentiate between unit tests and integration tests, we will need to differentiate them on a file level as well. This will result in cleaner code and cleaner interfaces as integration tests should only _know_ about external api.
+2. separation of test setup: we could theoretically easily deploy the integration test runner onto a custom machine
+3. by-passing integration tests: while not favorable, the advantage of this approach is that we can bypass the second stage if needed and merge directly into master without running the whole integration test suite but at least the unit test suite.
+
+[IMPORTANT]
+.Tests should be executed locally first
```

Comment:
I have to object to that statement a bit. Of course tests should be run locally, but our process should not depend on a dev ensuring locally all checks are met.

I understand that costs are coupled to using CI resources, but I'd rather follow an approach to reduce the required resources for running tests than to get dependent on a manual task. The proposed approach goes against a high degree of automation in my point of view.

Reply:
> I have to object to that statement a bit. Of course tests should be run locally, [...]

I think you are not objecting :) 

The CI is a backup system to ensure that the tests don't just run on your machine but also on others. It's not a replacement for running tests.

The problem is that each push to _try to make tests succeed_ on CI results in a new test runner to be spun up. This results in other actions having to wait because a new test runner cannot be allocated which slows down our overall development process.

Reply:
ok I see what you mean. so if I understand you correctly, that would mean that on bors r+ the heavy tests will be run and the action (merge) will only succeed if the tests succeed. Meaning we will not run into the issue I feared of potentially getting a broken main. correct?

Reply:
This is correct :)